### DISCUSSION TITLE
Security pin for connection instead only separate characteristics

### DISCUSSION THREAD

**Discussion by vovagorodok** [h2zero/NimBLE-Arduino#500] [Q&A]
Hi,

Here I found example secured characteristic:
https://github.com/h2zero/NimBLE-Arduino/blob/release/1.4/examples/NimBLE_Secure_Server/NimBLE_Secure_Server.ino
How to require pin before/for connection? Is here any example?

**Comment by h2zero** [comment]
In the onConnect callback you can start the security process by calling `NimBLEDevice::startSecurity()`.

**Reply by vovagorodok** [reply]
Done:
```
void ArduinoBleOTAClass::onConnect(NimBLEServer* pServer)
{
    NimBLEDevice::startSecurity(??);
}
```
but it requires `conn_id` param. Where to get it?

**Reply by h2zero** [reply]
You need to use the other onConnect callback:
```
    void onConnect(NimBLEServer* pServer, ble_gap_conn_desc* desc) {
        NimBLEDevice::startSecurity(desc->conn_handle);
    };
```

**Reply by vovagorodok** [reply]
It works! Thanks!
Do You know why pins with less than 6 digits are not accepted?

**Reply by h2zero** [reply]
The BLE specs say that it must be 6 digits.

**Reply by vovagorodok** [reply]
Can be more than 6, lats say u32::max according to BLE specs? Because currently more than 6 works

**Comment by vovagorodok** [comment]
One more question. How to implement yes/no button instead pin.
As I understand `BLE_HS_IO_KEYBOARD_ONLY` instead `BLE_HS_IO_DISPLAY_ONLY`
and implement some callback where allow if button pressed. Deprecated `bool NimBLESecurityCallbacks::onSecurityRequest()`?

**Reply by h2zero** [reply]
You would need to set the IO caps to `NimBLEDevice::setSecurityIOCap(BLE_HS_IO_DISPLAY_YESNO);`

**Reply by vovagorodok** [reply]
Deprecated `bool NimBLESecurityCallbacks::onSecurityRequest()` callback should be used, or there is new one?
